# minion-qcbench: Output

## Introduction

This document describes the output produced by the pipeline. 

The directories listed below will be created in the results directory after the pipeline has finished. All paths are relative to the top-level `results` directory.

## Pipeline overview

The pipeline is built using [Nextflow](https://www.nextflow.io/) and processes data using the following steps:

1. Filter reads using [`Chopper`](https://github.com/wdecoster/chopper) or [`PRINSEQ++`](https://github.com/Adrian-Cantu/PRINSEQ-plus-plus) by a minimum average phred score or leave the reads unfiltered
2. Assemble the preprocessed sequencing data using [`Flye`](https://github.com/fenderglass/Flye)
3. Calculate quality metrics for the assemblies [`QUAST`](https://github.com/ablab/quast)

### Chopper

<details markdown="1">
<summary>Output files</summary>

- `chopper/`
  - `*.fastq.gz`: FASTQ file with reads that passed the filters.

</details>

[`Chopper`](https://github.com/wdecoster/chopper), designed for long-read sequencing technologies like PacBio and ONT, filters and trims FASTQ files. In this project, the reads are filtered based on the average read quality, specified as a Phred score through the `--quality_scores` parameter in the pipeline.

### PRINSEQ++

<details markdown="1">
<summary>Output files</summary>

- `prinseqplusplus/`
  - `*_bad_out.fastq.gz`: FASTQ file with reads that didn't pass the filters.
  - `*_good_out.fastq.gz`: FASTQ file with reads that passed the filters.
  - `*.log`: Log file to keep track of parameters, errors, etc.

</details>

[`PRINSEQ++`](https://github.com/Adrian-Cantu/PRINSEQ-plus-plus) is a tool that can be used to filter, reformat, or trim sequence data. In this project, the reads are filtered based on the average read quality, specified as a Phred score through the `--quality_scores` parameter in the pipeline. Only those reads that pass the filter will be used in the pipeline for further processing.

### COPYFASTQ

<details markdown="1">
<summary>Output files</summary>

- `copyfastq/`
  - `*.fastq.gz`: Copy of the compressed FASTQ input file.

</details>

COPYFASTQ is a local module that creates a copy of a compressed FASTQ input file.

### Flye

<details markdown="1">
<summary>Output files</summary>

- `flye/`
    - `*.assembly.fasta.gz`: Final assembly.
    - `*.assembly_graph.{gfa|gv}.gz`: Final repeat graph.
    - `*.assembly_info.txt`: Extra information about contigs.
    - `*.flye.log`: Flye log file.
    - `*.params.json`: Flye parameters.

</details>

[`Flye`](https://github.com/fenderglass/Flye) is a de novo assembler designed for single-molecule sequencing reads, such as those generated by PacBio and Oxford Nanopore Technologies.

### QUAST

<details markdown="1">
<summary>Output files</summary>

- `quast/<sampleID>/`
    - `report.{txt|tsv|tex|pdf|html}`:	Assessment summary.
    - `icarus.html`: Icarus main menu with links to interactive viewers.
    - `contigs_reports/` (only if a reference genome is provided)
        - `misassemblies_report`: Detailed report on misassemblies.
        - `unaligned_report`:	Detailed report on unaligned and partially unaligned contigs.
<!-- TODO: Which files are actually produced?? -->
</details>

[`QUAST`](https://github.com/ablab/quast) (**QU**ality **AS**sessment **T**ool) evaluates genome assemblies by computing various quality metrics. For a detailed description of the output, please refer to the [`documentation`](https://quast.sourceforge.net/docs/manual.html#sec3).


### Pipeline information

<details markdown="1">
<summary>Output files</summary>

- `pipeline_info/`
  - Reports generated by Nextflow: `execution_report.html`, `execution_timeline.html`, `execution_trace.txt` and `pipeline_dag.dot`/`pipeline_dag.svg`.
  - Reports generated by the pipeline: `pipeline_report.html`, `pipeline_report.txt` and `software_versions.yml`. The `pipeline_report*` files will only be present if the `--email` / `--email_on_fail` parameter's are used when running the pipeline.
  - Reformatted samplesheet files used as input to the pipeline: `samplesheet.valid.csv`.
  - Parameters used by the pipeline run: `params.json`.

</details>

[Nextflow](https://www.nextflow.io/docs/latest/tracing.html) provides excellent functionality for generating various reports relevant to the running and execution of the pipeline. This will allow you to troubleshoot errors with the running of the pipeline, and also provide you with other information such as launch commands, run times and resource usage.
